src/bgw/scheduler.c

```diff
@@ -0,0 +1,402 @@
+/*
+ * This is a scheduler that takes background jobs and schedules them appropriately
+ *
+ * Limitations: For now the jobs are only loaded when the scheduler starts and are not
+ * updated if the jobs table changes
+ *
+ * TODO: right now jobs are not prioritized when slots available. Should prioritize in
+ * asc next_start order.
+ *
+*/
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <postmaster/bgworker.h>
+#include <storage/ipc.h>
+#include <storage/latch.h>
+#include <storage/lwlock.h>
+#include <storage/proc.h>
+#include <storage/shmem.h>
+#include <utils/jsonb.h>
+#include <utils/timestamp.h>
+#include <utils/snapmgr.h>
+#include <utils/memutils.h>
+#include <access/xact.h>
+#include <pgstat.h>
+
+#include "extension.h"
+#include "scheduler.h"
+#include "job.h"
+#include "job_stat.h"
+#include "version.h"
+#include "compat.h"
+
+#define MAX_TIMEOUT (5*1000L)
+#define SCHEDULER_APPNAME "TimescaleDB Background Worker Scheduler"
+
+TS_FUNCTION_INFO_V1(ts_bgw_scheduler_main);
+
+
+/* has to be global to shutdowns jobs on exit */
+static List *scheduled_jobs = NIL;
+
+typedef struct ScheduledBgwJob
+{
+	BgwJob	   *job;
+	TimestampTz next_start;
+	TimestampTz timeout_at;
+
+	bool		enabled;
+	BackgroundWorkerHandle *handle;
+} ScheduledBgwJob;
+
+static void scheduled_bgw_job_mark_stopped(ScheduledBgwJob *job);
+
+static inline void
+bgw_scheduler_on_postmaster_death(void)
+{
+	/* don't call exit hooks cause we want to bail out quickly */
+	on_exit_reset();
+	ereport(FATAL,
+			(errcode(ERRCODE_ADMIN_SHUTDOWN),
+			 errmsg("postmaster exited while timescaledb scheduler was working")));
+}
+static long
+get_timeout(TimestampTz by_time)
+{
+	long		timeout_sec = 5;
+	int			timeout_usec = 0;
+
+	if (TIMESTAMP_IS_NOBEGIN(by_time))
+		return 0;
+
+	TimestampDifference(GetCurrentTimestamp(), by_time, &timeout_sec, &timeout_usec);
+
+
+	if (timeout_sec <= 0 && timeout_usec <= 0)
+		return 0;
+
+	return timeout_sec * 1000L + timeout_usec / 1000;
+}
+
+static void
+scheduled_bgw_job_start_worker(ScheduledBgwJob *sjob)
+{
+
+	pid_t		pid;
+	BgwHandleStatus status;
+
+
+	Assert(sjob->enabled);
+
+	StartTransactionCommand();
+	ts_bgw_job_stat_mark_start(sjob->job->fd.id);
+	if (ts_bgw_job_has_timeout(sjob->job))
+		sjob->timeout_at = ts_bgw_job_timeout_at(sjob->job, GetCurrentTransactionStartTimestamp());
+	else
+		sjob->timeout_at = DT_NOEND;
+	CommitTransactionCommand();
+
+	elog(DEBUG1, "scheduling job %d \"%s\"", sjob->job->fd.id, NameStr(sjob->job->fd.job_name));
+
+
+	sjob->handle = ts_bgw_job_start_worker(sjob->job);
+	status = WaitForBackgroundWorkerStartup(sjob->handle, &pid);
+	switch (status)
+	{
+		case BGWH_POSTMASTER_DIED:
+			bgw_scheduler_on_postmaster_death();
+			break;
+		case BGWH_STARTED:
+			/* all good */
+			break;
+		case BGWH_STOPPED:
+			sjob->handle = NULL;
```

Comment:
note, we also null handle in `scheduled_bgw_job_mark_stopped`, if we need to do so here as well, we should comment why

Reply:
You are right. I cleaned up the logic here.